src/include/type/expression/abstract_expression.h

```diff
@@ -0,0 +1,191 @@
+#pragma once
+
+#include <memory>
+#include <vector>
+#include "common/hash_util.h"
+#include "common/json.h"
+#include "type/expression/expression_defs.h"
+#include "type/type_id.h"
+
+namespace terrier {
+namespace type {
+namespace expression {
+
+/**
+ * An abstract parser expression. Dumb and immutable.
+ */
+class AbstractExpression {
+ protected:
+  /**
+   * Instantiates a new abstract expression. Because these are logical expressions, everything should be known
+   * at the time of instantiation, i.e. the resulting object is immutable.
+   * @param expression_type what type of expression we have
+   * @param return_value_type the type of the expression's value
+   * @param children the list of children for this node
+   */
+  AbstractExpression(const ExpressionType expression_type, const TypeId return_value_type,
+                     const std::vector<AbstractExpression *> children)
+      : expression_type_(expression_type), return_value_type_(return_value_type) {
+    for (auto &child : children) {
+      children_.emplace_back(std::unique_ptr<AbstractExpression>(child));
+      if (child->is_nullable_) {
+        is_nullable_ = true;
+      }
+      if (child->has_parameter_) {
+        has_parameter_ = true;
+      }
+      if (child->has_subquery_) {
+        has_subquery_ = true;
+      }
+    }
+  }
+
+  /**
+   * (Convenience) Instantiates a new abstract expression for the 0-child case.
+   * @param expression_type what type of expression we have
+   * @param return_value_type the type of the expression's value
+   */
+  AbstractExpression(const ExpressionType expression_type, const TypeId return_value_type)
+      : expression_type_(expression_type), return_value_type_(return_value_type) {}
+
+  /**
+   * (Convenience) Instantiates a new abstract expression for the 1-child case.
+   * @param expression_type what type of expression we have
+   * @param return_value_type the type of the expression's value
+   * @param child child
+   */
+  AbstractExpression(const ExpressionType expression_type, const TypeId return_value_type, AbstractExpression *child)
+      : expression_type_(expression_type), return_value_type_(return_value_type) {
+    if (child->is_nullable_) {
+      is_nullable_ = true;
+    }
+    if (child->has_parameter_) {
+      has_parameter_ = true;
+    }
+    if (child->has_subquery_) {
+      has_subquery_ = true;
+    }
+    children_.emplace_back(std::unique_ptr<AbstractExpression>(child));
+  }
+
+  /**
+   * (Convenience) Instantiates a new abstract expression for the 2-child case.
+   * @param expression_type what type of expression we have
+   * @param return_value_type the type of the expression's value
+   * @param left left child
+   * @param right right child
+   */
+  AbstractExpression(const ExpressionType expression_type, const TypeId return_value_type, AbstractExpression *left,
```

Comment:
Why do we need multiple constructors to handle the cases for number of children, rather than being able to just use the constructor with the vector and just put 0,1,2, etc. children in the vector?